HADOOP-18177. Document prefetching architecture. (#4205)
apache/hadoop: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/prefetching.md

<!---
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License. See accompanying LICENSE file.
-->

# S3A Prefetching

This document explains the `S3PrefetchingInputStream` and the various components it uses.

> This input stream implements prefetching and caching to improve read performance of the input stream. A high level overview of this feature can also be found on [this](https://medium.com/pinterest-engineering/improving-efficiency-and-reducing-runtime-using-s3-read-optimization-b31da4b60fa0) blogpost.

Reviewer: Let's put the title in the link somehow for screenreaders.

```suggestion
This input stream implements prefetching and caching to improve read performance of the input stream. A high level overview of this feature was published in [Pinterest Engineering's blog post titled "Improving efficiency and reducing runtime using S3 read optimization"](https://medium.com/pinterest-engineering/improving-efficiency-and-reducing-runtime-using-s3-read-optimization-b31da4b60fa0).
```

> With prefetching, we divide the file into blocks of a fixed size (default is 8MB), associate buffers to these blocks, and then read data into these buffers asynchronously. We also potentially cache these blocks.

Reviewer: I might drop the number of references to the default in case we want to change it in the future.

Author: good point, have removed

> 

Reviewer 2: can you replace "we" in the docs to refer instead to the class/component which is doing the work

### Basic Concepts

> * **File** : A binary blob of data stored on some storage device.

Reviewer 2: should the doc use Object to refer to something in s3, to distinguish it from the local FS files?

* **Block :** A file is divided into a number of blocks. The default size of a block is 8MB, but can be configured. The size of the first n-1 blocks is same,  and the size of the last block may be same or smaller.

> * **Block based reading** : The granularity of read is one block. That is, we read an entire block and return or none at all. Multiple blocks may be read in parallel.

Reviewer: colon is inside the bold here but not for others

```suggestion
* **File**: A binary blob of data stored on some storage device.
* **Block**: A file is divided into a number of blocks. The default size of a block is 8MB, but can be configured. The size of the first n-1 blocks is same,  and the size of the last block may be same or smaller.
* **Block based reading**: The granularity of read is one block. That is, we read an entire block and return or none at all. Multiple blocks may be read in parallel.
```

### Configuring the stream

|Property    |Meaning    |Default    |
|---	|---	|---	|

> |fs.s3a.prefetch.enabled    |Enable the prefetch input stream    |TRUE |

Reviewer 2: 1. use backticks around the configuration names and values; all values must be valid if passed in as the config strings. they will be.
2. use `true` for true, rather than the capitalised value

> |fs.s3a.prefetch.block.size    |Size of a block    |8MB    |

Reviewer 2: 8MB isn't gong to be a valid value. `8M` should be

|fs.s3a.prefetch.block.count    |Number of blocks to prefetch    |8    |

> ### Key Components:

Reviewer: we can drop the trailing `:` in case its used in table of contents

`BlockData` - Holds information about the blocks in a file, such as:

* Number of blocks in the file
* Block size
* State of each block (initially all blocks have state *NOT_READY*). Other states are: Queued, Ready, Cached.

`BufferData` - Holds the buffer and additional information about it such as:

* The block number this buffer is for
* State of the buffer (Unknown, Blank, Prefetching, Caching, Ready, Done). Initial state of a buffer is blank.

`CachingBlockManager` - Implements reading data into the buffer, prefetching and caching.

`BufferPool` - Manages a fixed sized pool of buffers. It’s used by `CachingBlockManager` to acquire buffers.

`S3File` - Implements operations to interact with S3 such as opening and closing the input stream to the S3 file.

`S3Reader` - Implements reading from the stream opened by `S3File`. Reads from this input stream in blocks of 64KB.

`FilePosition` - Provides functionality related to tracking the position in the file. Also gives access to the current buffer in use.

`SingleFilePerBlockCache` - Responsible for caching blocks to the local file system. Each cache block is stored on the local disk as a separate file.

> ### Operation

Reviewer: What's the heading structure here?

Is it meant to be something like...

```
### Operation (Example Behavior?)

#### S3InMemoryInputStream

#### S3CachingInputStream

##### Sequential Reads

##### Random Reads
```

Nice examples!

### S3InMemoryInputStream

If we have a file with size 5MB, and block size = 8MB. Since file size is less than the block size, the `S3InMemoryInputStream` will be used.

If the caller makes the following read calls:

```
in.read(buffer, 0, 3MB);
in.read(buffer, 0, 2MB);
```

When the first read is issued, there is no buffer in use yet. We get the data in this file by calling the `ensureCurrentBuffer()` method, which ensures that a buffer with data is available to be read from.

The `ensureCurrentBuffer()` then:

* Reads data into a buffer by calling `S3Reader.read(ByteBuffer buffer, long offset, int size)`
*  `S3Reader` uses `S3File` to open an input stream to the S3 file by making a `getObject()` request with range as `(0, filesize)`.

> *  The S3Reader reads the entire file into the provided buffer, and once reading is complete closes the S3 stream and frees all underlying resources.

Reviewer: nit: extra spaces

```suggestion
* `S3Reader` uses `S3File` to open an input stream to the S3 file by making a `getObject()` request with range as `(0, filesize)`.
* The S3Reader reads the entire file into the provided buffer, and once reading is complete closes the S3 stream and frees all underlying resources.
```

* Now the entire file is in a buffer, set this data in `FilePosition` so it can be accessed by the input stream.

The read operation now just gets the required bytes from the buffer in `FilePosition`.

When the second read is issued, there is already a valid buffer which can be used. Don’t do anything else, just read the required bytes from this buffer.

### S3CachingInputStream

> [Image: image.png]

Reviewer: What image?

let's add alt text when embedding the image.

also, not sure what the best practice here for adding images to the docs. should they be stored in repo?

Reviewer 2: they go into hadoop-tools/hadoop-aws/src/site/resources/images ; other modules have examples of this

Author: sorry, I didn't mean to commit this. I was initially using the image in the blogpost: https://medium.com/pinterest-engineering/improving-efficiency-and-reducing-runtime-using-s3-read-optimization-b31da4b60fa0, but that diagram is very high level so I'm not sure if it adds much here. Do we think having that will help? Or do we think we need a lower level architecture diagram?

Now, if we have a file with size 40MB, and block size = 8MB. The `S3CachingInputStream` will be used.

### Sequential Reads

If the caller makes the following calls:

```
in.read(buffer, 0, 5MB)
in.read(buffer, 0, 8MB)
```

> For the first read call, there is no valid buffer yet. `ensureCurrentBuffer()` is called, and for the first read(), prefetch count is set as 1.

Reviewer 2: add backticks around 'read()'

When the second read is issued, there is already a valid buffer which can be used. 
Don’t do anything else, just read the required bytes from this buffer.

#### S3CachingInputStream

If there is a remote file with size 40MB and block size = 8MB, the `S3CachingInputStream` will be
used.

##### Sequential Reads

If the caller makes the following calls:

```
in.read(buffer, 0, 5MB)
in.read(buffer, 0, 8MB)
```

For the first read call, there is no valid buffer yet. 
`ensureCurrentBuffer()` is called, and for the first `read()`, prefetch count is set as 1.

The current block (block 0) is read synchronously, while the blocks to be prefetched (block 1) is
read asynchronously.

The `CachingBlockManager` is responsible for getting buffers from the buffer pool and reading data
into them. This process of acquiring the buffer pool works as follows:

* The buffer pool keeps a map of allocated buffers and a pool of available buffers. 
The size of this pool is = prefetch block count + 1. 
If the prefetch block count is 8, the buffer pool has a size of 9.
* If the pool is not yet at capacity, create a new buffer and add it to the pool.

> * If it’s at capacity, check if any buffers with state = done can be released. 

Reviewer: I noticed Yetus formats these `’` weirdly. Can we switch to standard single quote where-ever `’` occurs?

Releasing a buffer means removing it from allocated and returning it back to the pool of available 
buffers.
* If there are no buffers with state = done currently then nothing will be released, so retry the
  above step at a fixed interval a few times till a buffer becomes available.
* If after multiple retries there are still no available buffers, release a buffer in the ready state. 
The buffer for the block furthest from the current block is released.

Once a buffer has been acquired by `CachingBlockManager`, if the buffer is in a *READY* state, it is
returned. 
This means that data was already read into this buffer asynchronously by a prefetch. 

> If it’s state is *BLANK,* then data is read into it using 

Reviewer: we can move out the comma or drop it here

```suggestion
If it’s state is *BLANK* then data is read into it using 
```

# S3A Prefetching

This document explains the `S3PrefetchingInputStream` and the various components it uses.

This input stream implements prefetching and caching to improve read performance of the input

> stream. A high level overview of this feature was published in

Reviewer: For the line breaks, I believe Steve is suggesting each sentence start on a new line. If there's no empty line, Markdown will keep it part of the same paragraph.

So in this instance, we should move "A high level.." onto the next line. We probably want to do this throughout the file.

[Pinterest Engineering's blog post titled "Improving efficiency and reducing runtime using S3 read optimization"](https://medium.com/pinterest-engineering/improving-efficiency-and-reducing-runtime-using-s3-read-optimization-b31da4b60fa0)

> blogpost.

Reviewer: drop `blogpost` as it is already mentioned earlier

With prefetching, the input stream divides the remote file into blocks of a fixed size, associates
buffers to these blocks and then reads data into these buffers asynchronously. It also potentially
caches these blocks.

### Basic Concepts

* **Remote File**: A binary blob of data stored on some storage device.
* **Block**: A file is divided into a number of blocks. The size of the first n-1 blocks is same,
  and the size of the last block may be same or smaller.
* **Block based reading**: The granularity of read is one block. That is, either an entire block is
  read and returned or none at all. Multiple blocks may be read in parallel.

### Configuring the stream

|Property    |Meaning    |Default    |
|---	|---	|---	|
|fs.s3a.prefetch.enabled    |Enable the prefetch input stream    |`true` |
|fs.s3a.prefetch.block.size    |Size of a block    |`8M`    |

> |fs.s3a.prefetch.block.count    |Number of blocks to prefetch    |`8`    |

Reviewer: We want backticks around the configuration keys too.

```suggestion
|`fs.s3a.prefetch.enabled`    |Enable the prefetch input stream    |`true` |
|`fs.s3a.prefetch.block.size`    |Size of a block    |`8M`    |
|`fs.s3a.prefetch.block.count`    |Number of blocks to prefetch    |`8`    |
```

### Key Components

`S3PrefetchingInputStream` - When prefetching is enabled, S3AFileSystem will return an instance of
this class as the input stream. Depending on the remote file size, it will either use
the `S3InMemoryInputStream` or the `S3CachingInputStream` as the underlying input stream.

`S3InMemoryInputStream` - Underlying input stream used when the remote file size < configured block
size. Will read the entire remote file into memory.

`S3CachingInputStream` - Underlying input stream used when remote file size > configured block size.
Uses asynchronous prefetching of blocks and caching to improve performance.

`BlockData` - Holds information about the blocks in a remote file, such as:

* Number of blocks in the remote file
* Block size
* State of each block (initially all blocks have state *NOT_READY*). Other states are: Queued,
  Ready, Cached.

`BufferData` - Holds the buffer and additional information about it such as:

* The block number this buffer is for
* State of the buffer (Unknown, Blank, Prefetching, Caching, Ready, Done). Initial state of a buffer

>   is blank.

Reviewer: As an example for line breaks, we might instead have something like...

```suggestion
* State of the buffer (Unknown, Blank, Prefetching, Caching, Ready, Done).
  Initial state of a buffer is blank.
```

This way, we can comment or edit individual sentences - and also avoid breaking this sentence in half.

  pool is = prefetch block count + 1. If the prefetch block count is 8, the buffer pool has a size
  of 9.
* If the pool is not yet at capacity, create a new buffer and add it to the pool.
* If it’s at capacity, check if any buffers with state = done can be released. Releasing a buffer
  means removing it from allocated and returning it back to the pool of available buffers.
* If there are no buffers with state = done currently then nothing will be released, so retry the
  above step at a fixed interval a few times till a buffer becomes available.
* If after multiple retries there are still no available buffers, release a buffer in the ready
  state. The buffer for the block furthest from the current block is released.

Once a buffer has been acquired by `CachingBlockManager`, if the buffer is in a *READY* state, it is
returned. This means that data was already read into this buffer asynchronously by a prefetch. If
it’s state is *BLANK,* then data is read into it
using `S3Reader.read(ByteBuffer buffer, long offset, int size).`

For the second read call, `in.read(buffer, 0, 8MB)`, since the block sizes are of 8MB and only `5MB`

> of block 0 has been read so far, 3MB of the required data will be read from the current block 0.

Reviewer: Why only 5MB has backticks? Let's drop or add to all.
